Define what "safe"means (#347)
swiftlang/swift-book: TSPL.docc/LanguageGuide/TheBasics.md

Tuples enable you to create and pass around groupings of values.
You can use a tuple to return multiple values from a function as a single compound value.

Swift handles the absence of a value
using optional types.
Optionals say either “there *is* a value, which is *x*”
or “there *isn't* a value at all”.
Optionals ensure that code always
checks whether a value is missing before using the value,

> and non-optional values are guaranteed to never be missing.

Reviewer: ```suggestion
and nonoptional values are guaranteed to never be missing.
```

Author: TSPL uses the spelling "non-optional" throughout, per https://github.com/swiftlang/swift-book/blob/main/Style.md#non-optional

Swift is a safe language,

> which means it helps you find and fix several categories of bugs

Reviewer: ```suggestion
which means it makes it easier for you find and fix several categories of bugs
```

Reviewer: "Helps" sounds a little to anthropomorphic. What do you think of this change? Reword as necessary for accuracy.

Author: This rewrite is ok; changed.  I preferred "help" for brevity, and I don't think there's confusion here that Swift is helping you like how a person helps you, but I know we're taking a stronger stance against this kind of wording than when this chapter was first written.

Some of this intro was written with input from marketing, but this particular wording came from some revision in 2015 with Editorial, so I don't think we need to run the wording change by marketing for review.

as early as possible during the development process,

> and lets you guarantee that certain kinds of bugs can't happen.

Reviewer: ```suggestion
and makes it possible for you to prevent certain kinds of bugs from happening.
```

> Type safety helps you to be clear about

Reviewer: ```suggestion
Type safety enables you to be clear about
```

the types of values your code works with.
If part of your code requires a `String`,
type safety prevents you from passing it an `Int` by mistake.

> Data safety helps you work with only valid data,

Reviewer: ```suggestion
Data safety ensures that you work with valid data only,
```

## Type Safety and Type Inference

Every value in a Swift program has a type.
Every place you store a value ---
including constants, variables, and properties ---
also has a type,

> which you might write explicitly using a type annotation

Reviewer: ```suggestion
which you might write explicitly using a type annotation,
```

> or which might inferred from an initial value.

Reviewer: ```suggestion
or which might the system might infer from an initial value.
```

Reviewer: Is it "the system"? Reword as necessary for accuracy.

Author: It's not "the system" — I'll take a couple rewrite passes here, but this may also be example of where passive voice is clearer in TSPL in some places, because the actor is poorly defined and unimportant.  (See https://github.com/swiftlang/swift-book/blob/main/Style.md#passive-voice)

Other possibilities:

- "The language" doesn't work.  A programming language isn't an entity that can perform actions.
- "The compiler" is unfriendly.  While technically correct, TSPL specifically avoids using this term in the guide — and especially here, since it's the first chapter.  We do say "the compiler" in the reference
- "Swift" might work, but I'm not sure if it's actually any clearer.  This is similar to the example in the [TSPL style guide](https://github.com/swiftlang/swift-book/blob/main/Style.md#swift) for that term though, so it might be ok here too.

Every place in your code where you provide a value,
that value's type must match the place you use it.
For example,
if part of your code requires a `String`,
you can't pass it an `Int` by mistake.
This kind of checking makes Swift a *type-safe* language.

A type safe language encourages you to be clear about
the types of values your code works with.
Values of one type are never implicitly converted to another type.
However, some types can be explicitly converted.

> When building code

Reviewer: ```suggestion
When building code,
```

Swift also protects code against working with invalid memory.
This is known as *data safety* or *memory safety*
and includes the following requirements:

- Values are set before being read;
  code doesn't interact with uninitialized regions of memory.
- Arrays and buffers are accessed only at valid indexes,
  never out of bounds.
- Memory is accessed only during the value’s lifetime,
  preventing use-after-free errors.
- Access to memory overlaps only in provably safe ways;
  concurrent code doesn't create possible data races.

If you've worked in languages that don't provide these guarantees,
you may be familiar with some of the errors and bugs
named in the list above.
If you haven't encountered these issues, that's ok;
safe code in Swift avoids these problems.
For information about how Swift ensures you set initial values,
see <doc:Initialization>,
and for information about how Swift checks data safety in concurrent code,
see <doc:Concurrency>.

Sometimes you need to work outside of the bounds of safety ---
for example, because of limitations of the language or standard library ---
so Swift also provides unsafe versions of some APIs.

> When you use types or methods whose name includes the word "unsafe",

Reviewer 2: This is not the only word that announces unsafety. I know that there's also at least "unchecked" and "unmanaged", and occasionally there is nothing in particular. @glessard can confirm (if you want to elaborate on it).

Author: Reworded in 22f2822e.

  ```
-->

## Data Safety

In addition to the checks that prevent type mismatches,

> which are described above in <doc:TheBasics#Type-Safety-and-Type-Inference>,

Reviewer: ```suggestion
described above in <doc:TheBasics#Type-Safety-and-Type-Inference>,
```

Swift also protects code against working with invalid memory.

> This is known as *data safety* or *memory safety*

Reviewer: ```suggestion
This protection is known as *data safety* or *memory safety*
```

and includes the following requirements:

> - Values are set before being read;

Reviewer: The first three items would be clearer as active voice.

Author: I'll take a couple rewrite passes here, but this may also be example of where passive voice is clearer in TSPL in some places, because the actor is poorly defined and unimportant.  (See also https://github.com/swiftlang/swift-book/blob/main/Style.md#passive-voice)

Please see my above note about passive voice.  We could try rewriting together, but because there's no well-defined actor here, I think it may be difficult to make these active.  For example, the actor isn't "your code" or "you" because these guarantees also apply to code from libraries you use when they interact with your code.  The actor isn't "Swift" because the language and the compiler don't set or access values — code does that.

Trying to rewrite while carefully avoiding the need for an actor, just as an exercise:

> - Reading a value can only occur after writing to that value...
> - Access to arrays and buffers happens only at valid indexes...
> - Access to memory is allowed only during the value's lifetime...
> - Access to memory overlaps only in provably safe ways... [no change]

Maybe we'd be like the wording expressing these as requirements?  I avoided that phrasing because these are guarantees that the compiler makes to you, not things that you have to do.

> - A value must be set before any code can read from it...
> - Access to an array or buffer must use a valid index...
> - Access to memory must not occur before or after the value's lifetime...
> - Access to memory must not overlap, except in provably safe ways...

Thoughts?

  code doesn't interact with uninitialized regions of memory.
- Arrays and buffers are accessed only at valid indexes,
  never out of bounds.
- Memory is accessed only during the value’s lifetime,
  preventing use-after-free errors.
- Access to memory overlaps only in provably safe ways;
  concurrent code doesn't create possible data races.

If you've worked in languages that don't provide these guarantees,
you may be familiar with some of the errors and bugs
named in the list above.
If you haven't encountered these issues, that's ok;
safe code in Swift avoids these problems.
For information about how Swift ensures you set initial values,
see <doc:Initialization>,
and for information about how Swift checks data safety in concurrent code,
see <doc:Concurrency>.

Sometimes you need to work outside of the bounds of safety ---
for example, because of limitations of the language or standard library ---
so Swift also provides unsafe versions of some APIs.

> When you use types or methods whose name includes words like

Reviewer: ```suggestion
When you use types or methods with a name that includes words such as
```

"unsafe", "unchecked", or "unmanaged",
you take on the responsibility for safety.

Safe code in Swift can still encounter errors and unexpected failures,
which might stop the program's execution.

> Safety makes no guarantee about whether your code runs to completion.

Reviewer: ```suggestion
Safety doesn't ensure that your code runs to completion.
```
